tools/walletextension/ratelimiter/rate_limiter.go

```diff
@@ -52,56 +53,62 @@ func NewRateLimiter(threshold uint32, decay float64, maxScore uint32) *RateLimit
 	if rl.threshold == 0 {
 		return true
 	}
 
+	rl.mu.Lock()
+	defer rl.mu.Unlock()
+
+	rl.totalRequests++
 	now := time.Now()
 	userScore, exists := rl.users[userID]
 
 	if !exists {
 		// Create a new entry for the user if not exists
-		rl.users[userID] = Score{lastRequest: now, score: 0}
+		rl.users[userID] = Score{lastRequest: now, score: 0, concurrentRequests: 1}
+		return true
 	} else {
-		// Decay the score based on the time since the last request
-		timeSinceLastRequest := float64(now.Sub(userScore.lastRequest).Milliseconds())
-
-		// calculate the new score by subtracting the decay from the user's score
-		newScore := int64(userScore.score) - int64(timeSinceLastRequest*rl.decay)
-
-		// Ensure score does not become negative
-		if newScore < 0 {
-			newScore = 0
-		}
-
-		// Update user's score and last request time if new score is less than the current score
-		// use min value of new score and max score to ensure the score does not exceed the max score
-		// we check if the score changes before updating to avoid updating time and keeping the score same
-		if uint32(newScore) < userScore.score {
-			rl.users[userID] = Score{lastRequest: now, score: min(uint32(newScore), rl.maxScore)}
-		}
+		// Increase the number of concurrent requests by 1
+		rl.users[userID] = Score{lastRequest: now, score: userScore.score, concurrentRequests: userScore.concurrentRequests + 1}
 	}
 
-	// Check if the user's score is below the threshold
-	if rl.users[userID].score > rl.threshold {
+	// Check if the user's score is below the threshold (taking into account the decay)
+	timeSinceLastRequest := float64(now.Sub(userScore.lastRequest).Milliseconds())
+	newScore := int64(userScore.score) - int64(timeSinceLastRequest*rl.decay)
+	if newScore > int64(rl.threshold) || userScore.concurrentRequests > rl.maxConcurrentRequests {
 		rl.rateLimitedRequests++
 		return false
 	}
 	return true
 }
 
-// UpdateScore updates the score of the user based on the additional score (time taken to process the request)
-func (rl *RateLimiter) UpdateScore(userID common.Address, additionalScore uint32) {
+// UpdateScore updates the score of the user based on the execution duration of the request.
+func (rl *RateLimiter) UpdateScore(userID common.Address, executionDuration uint32) {
+	// If the threshold is 0,rate limiting is disabled and there is no need for updating the score
+	if rl.threshold == 0 {
+		return
+	}
 	rl.mu.Lock()
 	defer rl.mu.Unlock()
-	newScore := rl.users[userID].score + additionalScore
-	rl.users[userID] = Score{lastRequest: time.Now(), score: newScore}
+
+	// Decay the score of the user based on the time since the last request
+	now := time.Now()
+	userScore, exists := rl.users[userID]
+	scoreDecay := uint32(0)
+	// if user exists decay the score based on the time since the last request
+	if exists {
+		// Decay the score based on the time since the last request and the decay rate
+		timeSinceLastRequest := float64(now.Sub(userScore.lastRequest).Milliseconds())
```

Comment:
I don't quite understand why we need to decay the score here. It's hard to reason about what's going on, as that responsibility is now in 2 places.

Reply:
We decay it only here in `UpdateScore`. 
But we need to calculate it also in `Allow` function to compare it to the threshold is this now what you meant in this comment: https://github.com/ten-protocol/go-ten/pull/1942#discussion_r1632877645 ?